tracer/src/Datadog.Trace/IAST/Aspects/System/StringAspects.cs

```diff
@@ -272,4 +272,289 @@ public static char[] ToCharArray(string target, int startIndex, int length)
+    /// </summary>
+    /// <param name="separator"> sparator </param>
+    /// <param name="values"> values to join </param>
+    /// <param name="startIndex"> start index </param>
+    /// <param name="count"> number of elemnts to join </param>
+    /// <returns> Join result </returns>
+    [AspectMethodReplace("System.String::Join(System.Char,System.String[],System.Int32,System.Int32)")]
+    public static string Join(char separator, string[] values, int startIndex, int count)
+    {
+        return OnStringJoin(string.Join(separator, values, startIndex, count), values, startIndex, count);
+    }
+
+    /// <summary>
+    /// String.Join aspect
+    /// </summary>
+    /// <param name="separator"> sparator </param>
+    /// <param name="values"> values to join </param>
+    /// <returns> Join result </returns>
+    [AspectMethodReplace("System.String::Join(System.Char,System.Collections.Generic.IEnumerable`1<!!0>)")]
+    public static string Join(char separator, IEnumerable values)
+    {
+        return Join(separator.ToString(), values);
+    }
+#endif
+
+    /// <summary>
+    /// String.Join aspect
+    /// </summary>
+    /// <param name="separator"> sparator </param>
+    /// <param name="values"> values to join </param>
+    /// <returns> Join result </returns>
+    [AspectMethodReplace("System.String::Join(System.String,System.Object[])")]
+    public static string Join(string separator, object[] values)
+    {
+        return OnStringJoin(string.Join(separator, values), separator, values);
+    }
+
+    /// <summary>
+    /// String.Join aspect
+    /// </summary>
+    /// <param name="separator"> sparator </param>
+    /// <param name="values"> values to join </param>
+    /// <returns> Join result </returns>
+    [AspectMethodReplace("System.String::Join(System.String,System.String[])")]
+    public static string Join(string separator, string[] values)
+    {
+        return OnStringJoin(string.Join(separator, values), separator, values);
+    }
+
+    /// <summary>
+    /// String.Join aspect
+    /// </summary>
+    /// <param name="separator"> sparator </param>
+    /// <param name="values"> values to join </param>
+    /// <returns> Join result </returns>
+    [AspectMethodReplace("System.String::Join(System.String,System.Collections.Generic.IEnumerable`1<System.String>)")]
+    public static string Join(string separator, IEnumerable values)
+    {
+        if (values is null)
+        {
+            return OnStringJoin(string.Join(separator, values), separator, null);
+        }
+
+        var valuesConverted = values as IEnumerable<string>;
+        if (valuesConverted != null)
+        {
+            return OnStringJoin(string.Join(separator, valuesConverted), separator, valuesConverted);
+        }
+        else
+        {
+            // This should neve happen
+            Log.Warning("Could not taint the string.join call in System.String::Join(System.String,System.Collections.Generic.IEnumerable`1<System.String>)");
```

Comment:
Should this be a debug log instead even though it should never happen?

Reply:
I like it being a warning because if we ever reach this situation, we should be aware of it and investigate it. On the other hand, reaching this situation is something that we should investigate, but it is not going to cause any problem to the client other that some value in a very strange situation would not be tainted. What do you think?